pkg/sql/catalog/hydratedtables/hydratedcache.go

```diff
@@ -0,0 +1,327 @@
+
+var cacheKeySyncPool = sync.Pool{
+	New: func() interface{} { return new(cacheKey) },
+}
+
+func newCacheKey(table catalog.TableDescriptor) *cacheKey {
+	k := cacheKeySyncPool.Get().(*cacheKey)
+	*k = cacheKey{
+		ID:      table.GetID(),
+		Version: table.GetVersion(),
+	}
+	return k
+}
+
+func putCacheKey(k *cacheKey) {
+	*k = cacheKey{}
+	cacheKeySyncPool.Put(k)
+}
+
+// GetHydratedTableDescriptor returns an ImmutableCopy with the types
+// hydrated. It may use a cached copy but all of the relevant type descriptors
+// will be retrieved via the TypeDescriptorResolver.
+func (c *Cache) GetHydratedTableDescriptor(
+	ctx context.Context, table *tabledesc.Immutable, res catalog.TypeDescriptorResolver,
+) (_ *tabledesc.Immutable, canUseCached bool, err error) {
+	if table.IsModified() {
+		return nil, false, nil
+	}
+	if !table.ContainsUserDefinedTypes() {
+		return table, true, nil
+	}
+
+	// TODO(ajwerner): This cache may thrash a bit right when a version of a type
+	// changes as different callers oscillate evicting the old and new versions of
+	// that type. It should converge rapidly but nevertheless, a finer-granularity
+	// cache could mitigate the problem. The idea would be to cache all tuples
+	// of table and type versions and then check if what we get from the resolver
+	// matches any of them.
+	var groupKey string // used as a proxy for cache hit
+	defer func() {
+		if err != nil || !canUseCached {
+			return
+		}
+		if groupKey == "" {
+			c.metrics.Hits.Inc(1)
+		} else {
+			c.metrics.Misses.Inc(1)
+		}
+	}()
+	key := newCacheKey(table)
+	defer func() {
+		if key != nil {
+			putCacheKey(key)
+		}
+	}()
+	for {
+		cached, ok := c.getHydratedTableDescriptorFromCache(ctx, key)
+		if ok {
+			canUse, hasModified, err := canUseCachedDescriptor(ctx, cached, res)
+			if err != nil || hasModified {
+				return nil, false, err
+			}
+			if canUse {
+				return cached.tableDesc, true, nil
+			}
+		}
+		if groupKey == "" {
+			groupKey = fmt.Sprintf("%d@%d", key.ID, key.Version)
+		}
+		// Now we want to lock the cache and populate the descriptors.
+		// Only the calling goroutine can actually use the result directly.
+		// Furthermore, if an error is encountered, only the calling goroutine
+		// should return it.
+		var called bool
+		res, _, err := c.g.Do(groupKey, func() (interface{}, error) {
+			called = true
+			cachedRes := cachedTypeDescriptorResolver{
+				underlying: res,
+				cache:      map[descpb.ID]*cachedType{},
+			}
+			descBase := protoutil.Clone(table.TableDesc()).(*descpb.TableDescriptor)
+			if err := typedesc.HydrateTypesInTableDescriptor(ctx, descBase, &cachedRes); err != nil {
+				return nil, err
+			}
+			hydrated := tabledesc.NewImmutable(*descBase)
+
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			if !cachedRes.haveModified {
+				c.mu.cache.Add(key, &hydratedTableDescriptor{
+					tableDesc: hydrated,
+					typeDescs: cachedRes.types,
+				})
+			}
+			key = nil // prevent the key from being put back in the pool
```

Comment:
why does this matter

Reply:
because it's in the data structure, commented